#!/usr/bin/perl
use Getopt::Long;

GetOptions (\%opt,"ortholog:s","inf1:s","inf2:s","project:s","help");


my $help=<<USAGE;
Bar plot of Expression level comparison between ortholog genes based on different categray TE+-/Me+-.
perl $0 --orth ../input/OB-OS.orth --inf1 ./Distance2Expression/0.1/1k/OBa_Me.gene.closestTE.inf --inf2 ./Distance2Expression/0.1/1k/rice_Me.gene.closestTE.inf --project orth_OBa_rice > log 2> log2 &
--ortholog: ortholog table, the species cresponding to inf1 should be in the first colume
--inf1: gene information table with gene name, RPKM, closest TE distance, TE ID, TE methylation status, which generated by Distance2RPKM.pl
Gene    RPKM    Repeat  Distance        Methylation
--inf2: same with inf1, the other species.
--project: 
USAGE


if ($opt{help} or keys %opt < 1){
    print "$help\n";
    exit();
}

my $methylevel=0.1; ### methylation cutoff
my $TEdist=100; ### closest TE cutoff
my $refinf1=geneinf($opt{inf1});
my $refinf2=geneinf($opt{inf2});

& orthcompare($opt{ortholog},$refinf1,$refinf2,$methylevel,$TEdist);

sub orthcompare
{
my ($orth,$refinf1,$refinf2,$methylevel,$TEdist)=@_;
my %hash;
my %spe1;
my %spe2;
my %ort;
open IN, "$orth" or die "$!";
while(<IN>){
   chomp $_;
   my @unit=split("\t",$_);
   if (exists $refinf1->{$unit[0]} and exists $refinf2->{$unit[1]}){
       my $diff=$refinf1->{$unit[0]}->[0]-$refinf2->{$unit[1]}->[0];
       my $exp1=$refinf1->{$unit[0]}->[0];
       my $exp2=$refinf2->{$unit[1]}->[0];
       my $tmp1="$refinf1->{$unit[0]}->[0]\t$refinf1->{$unit[0]}->[1]\t$refinf1->{$unit[0]}->[2]\t$refinf1->{$unit[0]}->[3]";
       my $tmp2="$refinf2->{$unit[1]}->[0]\t$refinf2->{$unit[1]}->[1]\t$refinf2->{$unit[1]}->[2]\t$refinf2->{$unit[1]}->[3]";
       my $ort="$unit[0]\t$tmp1\t$unit[1]\t$tmp2";
       if ($refinf1->{$unit[0]}->[2] <= $TEdist and $refinf1->{$unit[0]}->[2] > 0){
           if ($refinf2->{$unit[1]}->[2] <= $TEdist and $refinf2->{$unit[1]}->[2] > 0){
               push (@{$hash{1}},$diff); 
               push (@{$ort{1}},$ort);
               push (@{$spe1{1}},$exp1);
               push (@{$spe2{1}},$exp2);
           }else{
               if ($refinf1->{$unit[0]}->[3] >= $methylevel){
                   push (@{$hash{3}},$diff);
                   push (@{$ort{3}},$ort);
                   push (@{$spe1{3}},$exp1);
                   push (@{$spe2{3}},$exp2);
               }else{
                   push (@{$hash{2}},$diff);
                   push (@{$ort{2}},$ort);
                   push (@{$spe1{2}},$exp1);
                   push (@{$spe2{2}},$exp2);
               }
           }
       }else{
           if ($refinf2->{$unit[1]}->[2] > $TEdist){
               push (@{$hash{0}},$diff);
               push (@{$ort{0}},$ort);
               push (@{$spe1{0}},$exp1);
               push (@{$spe2{0}},$exp2);
           }else{
               if ($refinf2->{$unit[1]}->[3] >= $methylevel){
                   push (@{$hash{5}},$diff);
                   push (@{$ort{5}},$ort);
                   push (@{$spe1{5}},$exp1);
                   push (@{$spe2{5}},$exp2);
               }else{
                   push (@{$hash{4}},$diff);
                   push (@{$ort{4}},$ort);
                   push (@{$spe1{4}},$exp1);
                   push (@{$spe2{4}},$exp2);
               }
           }
       }
       #TE in gene intron/exon
       #if ($refinf1->{$unit[0]}->[2] == 0 and $refinf2->{$unit[1]}->[2] == 0){
       #    push (@{$hash{6}},$diff);
       #}
   }
}
close IN;

my %class=(
   "0" => "TE-\|TE-",
   "1" => "TE+\|TE+",
   "2" => "mC-\|TE-",
   "3" => "mC+\|TE-",
   "4" => "TE-\|mC-",
   "5" => "TE-\|mC+"
   #"6" => "TE0\|TE0"
);

open INF, ">$opt{project}.ortholog.inf" or die "$!";
open OUT, ">$opt{project}.4r" or die "$!";
foreach(sort {$a <=> $b} keys %hash){
    print "$_\t$class{$_}\n";
    my $id=$_;
    foreach(@{$ort{$id}}){
         print INF "$class{$id}\t$_\n";
    }
    my ($mean,$se,$number)=ci($hash{$id},$id);
    my ($mean1,$se1,$number1)=ci($spe1{$id},$id);
    my ($mean2,$se2,$number2)=ci($spe2{$id},$id);
    my $pvalue=ttest($spe1{$id},$spe2{$id},$id);
    my $cilow=$mean-$se;
    my $cihigh=$mean+$se;
    my $cilow1=$mean1-$se1;
    my $cihigh1=$mean1+$se1;
    my $cilow2=$mean2-$se2;
    my $cihigh2=$mean2+$se2;
    print OUT "$class{$id}\t$mean\t$cilow\t$cihigh\t$number\t$mean1\t$cilow1\t$cihigh1\t$number1\t$mean2\t$cilow2\t$cihigh2\t$number2\t$pvalue\n";
}
close OUT;
close INF;

open OUT, ">$opt{project}.r" or die "$!";
print OUT <<"END.";
read.table("$opt{project}.4r") -> x
pdf("$opt{project}.pdf")
library("gplots")
barplot2(x[,2],plot.ci=TRUE,ci.l=x[,3],ci.u=x[,4],names.arg=x[,1],ylab="Normalized expression difference")
mean <- rbind(x[,6],x[,10])
cilow <- rbind(x[,7],x[,11])
cihigh <- rbind(x[,8],x[,12])
barplot2(mean,ylim=c(0,1.5),beside = TRUE,plot.ci=TRUE,col = c("gray38", "gray57"),names.arg=x[,1],ci.l=cilow,ci.u=cihigh,ylab="Normalized expression level")
savefont <- par(font=3)
legend(1,1.4,c("O. brachyantha","O. sativa"),lty=1,lwd=12,col=c("gray38", "gray57"))
abline(h=0)
dev.off()
END.
close OUT;

system ("cat $opt{project}.r | R --vanilla --slave");
} ### end of sub orthcompare 

#############################################
sub ttest
{
my ($num1,$num2,$order)=@_;
open OUT2, ">$opt{project}.$order.ttest.4r" or die "$!";
for(my $i=0;$i<@$num1;$i++){
   print OUT2 "$num1->[$i]\t$num2->[$i]\n";
}
close OUT2;

open OUT1, ">$opt{project}.$order.ttest.r" or die "$!";
print OUT1 <<"END.";
read.table("$opt{project}.$order.ttest.4r") -> x
t <- t.test(x[,1],x[,2],paired=TRUE)
p <-t\$p.value
p
END.
close OUT1;
my $tmp=`cat $opt{project}.$order.ttest.r | R --vanilla --slave`;
my @tmp=split(" ",$tmp);
my $pvalue=$tmp[1];
print "Pvalue\t$pvalue\n";
return $pvalue;
}


sub ci
{
my ($num,$order)=@_;
my $loop=0;
my $total;
my $add_square;
open OUT1, ">$opt{project}.$order.4r" or die "$!";
foreach  (@$num) {
        next if ($_ eq "NA");
        #my $temp=log($_);
        #print "$_\t$temp\n";
        print OUT1 "$_\n";
        my $temp=$_;
        $total+=$temp;
        $add_square+=$temp*$temp;
        $loop++;
}
close OUT1;
=pod
open OUT1, ">$opt{project}.$order.r" or die "$!";
print OUT1 <<"END.";
read.table("$opt{project}.$order.4r") -> x
pdf("$opt{project}.$order.pdf")
plot(density(x[,1]),xlim=c(0,20));
dev.off()
END.
close OUT1;
system ("cat $opt{project}.$order.r | R --vanilla --slave");
=cut

my $number=$loop;
my $mean=$total/$number;
my $SD=sqrt( ($add_square-$total*$total/$number)/ ($number-1) );
my $se=1.96*$SD/sqrt($number);
return ($mean,$se,$number);
}

sub log2 {
    my ($n) = shift;
    return log($n)/log(2);
}



sub geneinf
{
###Gene    RPKM    Repeat  Distance        Methylation
my ($inf)=@_;
my %hash;
open IN, "$inf" or die "$!";
<IN>;
while(<IN>){
    chomp $_;
    my @unit=split("\t",$_);
    next if ($unit[1] eq "NA" or $unit[4] eq "NA");
    $hash{$unit[0]}=[$unit[1],$unit[2],$unit[3],$unit[4]];
}
close IN;
return \%hash;
}
